numba/array_intern.py

```diff
@@ -0,0 +1,112 @@
+from __future__ import print_function, division, absolute_import
+import numpy as np
+from numba import cgutils, ir, types, rewrites
+from numba.typing.templates import FunctionTemplate
+
+
+@rewrites.register_rewrite
+class InternArrayAlloc(rewrites.Rewrite):
+    def __init__(self, *args, **kwargs):
+        super(InternArrayAlloc, self).__init__(*args, **kwargs)
+        self._block = None
+        self._matches = None
+        self._enabled = self.pipeline.targetctx.enable_array_intern
+        if self._enabled:
+            # Install a lowering hook if we are using this rewrite.
+            special_ops = self.pipeline.targetctx.special_ops
+            if 'interned_alloc' not in special_ops:
+                special_ops['interned_alloc'] = _lower_interned_alloc
+
+    def match(self, label, block, typemap, calltypes):
+        if not self._enabled:
+            return False
+        # Match only the first block
+        if label == 0 and block.body and len(calltypes) > 1:
```

Comment:
I'm wondering if we can find something better than `label == 0` (which sounds a bit obscure) to test for this.